message.go

```diff
@@ -4,54 +4,246 @@ import (
 	"errors"
 )
 
-type message struct {
-	method          string
-	seqno           seqNumber
-	res             interface{}
-	err             interface{}
-	remainingFields int
-	decodeSlots     []interface{}
-}
-
-func decodeIntoMessage(dec decoder, m *message) error {
-	for _, s := range m.decodeSlots {
-		if err := decodeMessage(dec, m, s); err != nil {
-			return err
+type RPCMessage interface {
+	Type() MethodType
+	Decode(l int, d decoder, p *protocolHandler, cc *callContainer) error
+	RPCData
+}
+
+type RPCData interface {
```

Comment:
I feel like this interface is a union of a few separate interfaces, as evidenced by the necessity to have BasicRPCData. I think you can leave the common functions in here, but then have separate interfaces that have whatever combinations of ResponseCh(), Err(), Res() you need, and then the caller can just do checked type conversions when they need those fields. That way, we don't need BasicRPCData.